File: src/RealtimeServer/scriptureforge/models/sf-project-user-config.ts
import { PROJECT_DATA_INDEX_PATHS, ProjectData } from '../../common/models/project-data';
import { EditorTabPersistData } from './editor-tab-persist-data';
import { LynxUserConfig } from './lynx-config';
import { LynxInsightUserData } from './lynx-insight-user-data';

export const SF_PROJECT_USER_CONFIGS_COLLECTION = 'sf_project_user_configs';
export const SF_PROJECT_USER_CONFIG_INDEX_PATHS: string[] = PROJECT_DATA_INDEX_PATHS;

export function getSFProjectUserConfigDocId(projectId: string, userId: string): string {
  return `${projectId}:${userId}`;
}

export interface SFProjectUserConfig extends ProjectData {
  selectedTask?: string;
  selectedQuestionRef?: string;
  selectedBookNum?: number;
  selectedChapterNum?: number;
  selectedBiblicalTermsCategory?: string;
  selectedBiblicalTermsFilter?: string;
  isTargetTextRight: boolean;
  confidenceThreshold: number;
  biblicalTermsEnabled?: boolean;
  transliterateBiblicalTerms: boolean;
  translationSuggestionsEnabled: boolean;
  numSuggestions: number;
  selectedSegment: string;
  selectedSegmentChecksum?: number;
  noteRefsRead: string[];
  questionRefsRead: string[];
  answerRefsRead: string[];
  commentRefsRead: string[];
  editorTabsOpen: EditorTabPersistData[];
  lynxInsightState: LynxInsightUserData;
  lynxUserConfig: LynxUserConfig;
}

File: src/SIL.XForge.Scripture/Models/SFProjectUserConfig.cs
using System;
using System.Collections.Generic;
using SIL.XForge.Models;

namespace SIL.XForge.Scripture.Models;

public class SFProjectUserConfig : ProjectData
{
    public static string GetDocId(string projectId, string userId) => $"{projectId}:{userId}";

    public string SelectedTask { get; set; }
    public int? SelectedBookNum { get; set; }
    public int? SelectedChapterNum { get; set; }

    // translate
    public bool IsTargetTextRight { get; set; } = true;
    public double ConfidenceThreshold { get; set; } = 0.2;
    public int NumSuggestions { get; set; } = 1;
    public bool TranslationSuggestionsEnabled { get; set; } = true;
    public string SelectedSegment { get; set; } = "";
    public int? SelectedSegmentChecksum { get; set; }

    /// <summary>Ids of notes (not threads) which the user has read.</summary>
    public List<string> NoteRefsRead { get; set; } = [];

    // checking
    public List<string> QuestionRefsRead { get; set; } = [];
    public List<string> AnswerRefsRead { get; set; } = [];
    public List<string> CommentRefsRead { get; set; } = [];
    public List<EditorTabPersistData> EditorTabsOpen { get; set; } = [];
    public LynxInsightUserData? LynxInsightState { get; set; }
    public string? SelectedQuestionRef { get; set; }

    [Obsolete("For backwards compatibility with older frontend clients. Deprecated September 2024.")]
    public bool? BiblicalTermsEnabled { get; set; }
    public bool TransliterateBiblicalTerms { get; set; }
    public string? SelectedBiblicalTermsCategory { get; set; }
    public string? SelectedBiblicalTermsFilter { get; set; }
}
